SignalServiceKit/src/Messages/OWSReadReceiptManager.m

```diff
@@ -0,0 +1,191 @@
+}
+
+- (void)databaseViewRegistrationComplete
+{
+    [self scheduleProcessing];
+}
+
+// Schedules a processing pass, unless one is already scheduled.
+- (void)scheduleProcessing
+{
+    DispatchMainThreadSafe(^{
+        @synchronized(self)
+        {
+            if ([TSDatabaseView hasPendingViewRegistrations]) {
+                return;
+            }
+            if (self.isProcessing) {
+                return;
+            }
+
+            self.isProcessing = YES;
+
+            // Process read receipts every N seconds.
+            //
+            // We want a value high enough to allow us to effectively deduplicate,
+            // read receipts without being so high that we risk not sending read
+            // receipts due to app exit.
+            const CGFloat kProcessingFrequencySeconds = 3.f;
+            dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(kProcessingFrequencySeconds * NSEC_PER_SEC)),
+                dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0),
+                ^{
+                    [self process];
+                });
+        }
+    });
+}
+
+- (void)process
+{
+    @synchronized(self)
+    {
+        self.isProcessing = NO;
+
+        NSArray<OWSReadReceipt *> *readReceiptsToSend = [[self.toLinkedDevicesReadReceiptMap allValues] copy];
+        [self.toLinkedDevicesReadReceiptMap removeAllObjects];
+        if (readReceiptsToSend.count > 0) {
+            OWSReadReceiptsMessage *message = [[OWSReadReceiptsMessage alloc] initWithReadReceipts:readReceiptsToSend];
+
+            dispatch_async(dispatch_get_main_queue(), ^{
+                [self.messageSender sendMessage:message
+                    success:^{
+                        DDLogInfo(@"%@ Successfully sent %zd read receipt to linked devices.",
+                            self.tag,
+                            readReceiptsToSend.count);
+                    }
+                    failure:^(NSError *error) {
+                        DDLogError(@"%@ Failed to send read receipt to linked devices with error: %@", self.tag, error);
+                    }];
+            });
+        }
+    }
+}
+
+- (void)messageWasReadLocally:(TSIncomingMessage *)message;
+{
+    @synchronized(self)
+    {
+        NSString *threadUniqueId = message.thread.uniqueId;
+        OWSAssert(threadUniqueId.length > 0);
+
+        // Only groupthread sets authorId, thus this crappy code.
+        // TODO Refactor so that ALL incoming messages have an authorId.
+        NSString *messageAuthorId;
+        if (message.authorId) {
+            // Group Thread
+            messageAuthorId = message.authorId;
+        } else {
+            // Contact Thread
+            messageAuthorId = [TSContactThread contactIdFromThreadId:message.uniqueThreadId];
+        }
+        OWSAssert(messageAuthorId.length > 0);
+
+        OWSReadReceipt *newReadReceipt =
+            [[OWSReadReceipt alloc] initWithSenderId:messageAuthorId timestamp:message.timestamp];
```

Comment:
Clarifying question: 

So, at the individual OWSReadReceipt model level:

e.g. `[[OWSReadReceipt alloc] initWithSenderId:@"AlicePhoneNumber" timestamp:123]`

Could represent either:
- I read a message from Alice, with timestamp 123, on my linked device
- or that Alice read a message from me sent at 123

Presumably the distinction will be made just by whether the read receipt is in the `toLinkedDevicesReadReceiptMap` and a forthcoming `fromOtherUsersReadReceiptMap` (e.g.).

Is that right?

Technically that's enough information, but for sanity we may want to include an explicit type (incoming vs. outgoing) so that we can make asserts.

Take it or leave it - we can revisit it later if it seems like it could be an issue.

Reply:
There's four kind of read receipts:

* Sent to linked device.
* Received from linked device.
* Sent to message sender.
* Received from message recipient.

First, `OWSReadReceipt` is used to represent the first two, not the last two.  

Secondly, `OWSReadReceipt` isn't a new class - I've just streamlined the logic that uses it.  I was tempted to rename/rework it, but I didn't want to bother with the work/possible bugs that would come from that kind of migration.